core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/common/accessor/gizmo/GizmoFieldHandler.java

```diff
@@ -0,0 +1,82 @@
+package org.optaplanner.core.impl.domain.common.accessor.gizmo;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.util.function.Consumer;
+
+import io.quarkus.gizmo.BytecodeCreator;
+import io.quarkus.gizmo.FieldDescriptor;
+import io.quarkus.gizmo.MethodDescriptor;
+import io.quarkus.gizmo.ResultHandle;
+
+final class GizmoFieldHandler implements GizmoMemberHandler {
+
+    private final Class<?> declaringClass;
+    private final Field field;
+    private final FieldDescriptor fieldDescriptor;
+    private final boolean canBeWritten;
+
+    GizmoFieldHandler(Class<?> declaringClass, Field field, FieldDescriptor fieldDescriptor, boolean canBeWritten) {
+        this.declaringClass = declaringClass;
+        this.field = field;
+        this.fieldDescriptor = fieldDescriptor;
+        this.canBeWritten = canBeWritten;
+    }
+
+    @Override
+    public void whenIsField(Consumer<FieldDescriptor> fieldDescriptorConsumer) {
+        fieldDescriptorConsumer.accept(fieldDescriptor);
+    }
+
+    @Override
+    public void whenIsMethod(Consumer<MethodDescriptor> methodDescriptorConsumer) {
+        // Do nothing.
+    }
+
+    @Override
+    public ResultHandle readMemberValue(BytecodeCreator bytecodeCreator, ResultHandle thisObj) {
+        return bytecodeCreator.readInstanceField(fieldDescriptor, thisObj);
+    }
+
+    @Override
+    public boolean writeMemberValue(MethodDescriptor setter, BytecodeCreator bytecodeCreator, ResultHandle thisObj,
+            ResultHandle newValue) {
+        try {
+            bytecodeCreator.writeInstanceField(fieldDescriptor, thisObj, newValue);
+            return true;
+        } catch (Exception ex) {
+            if (canBeWritten) {
+                throw new IllegalStateException("Cannot write field (" + field + ") of class (" + declaringClass + ").", ex);
+            } else {
+                return false;
+            }
+        }
```

Comment:
This is kinda strange. In order to make all tests pass, I had to attempt to write in all cases, and then only throw the exception if the write should have succeeded.

Instead, I'd expect not to have to do the write if I know it should fail.

Reply:
Hmm. I explicitly check if the field is final. Are you saying that Quarkus makes the field non-final, while the field is still reporting itself to be final?

Reply:
Yes; this code is called BEFORE the transformation happens, hence it will report as final.

Reply:
Ah, ok, so that explain the issue very well. Gap in my understanding, as usual.